### PR TITLE
1947-data-page-component

### PR DESCRIPTION
**Pull request by mlbrgl**
Adds the necessary architecture for the first ~20 data pages prototype, along with the first third of data pages visual components (see [figma](https://www.figma.com/file/hGrJi0P2n2TLsUFO2ispxe/%5BProject%5D-Data-Pages?node-id=1-19057&t=2xJDhcYuEBp25Rqb-0)).

The ambition of this PR is to get this first third in a state where it could be used for publishing - incomplete but technically stable - data pages (pending mobile version).

Screenshot 2023-04-03 at 20 52 48

Editing flow:
- edit [JSON on staging branch](https://github.com/owid/owid-content/blob/staging/datapages/538918.json)
- "Pull updates from Github" in [Explorers admin](https://exemplars.owid.cloud/admin/explorers#)
- refresh [admin-only datapage view](https://exemplars.owid.cloud/admin/grapher/co-emissions-per-capita) on https://exemplars.owid.cloud/admin

Resources:
- #1947 
- [Figma](https://www.figma.com/file/hGrJi0P2n2TLsUFO2ispxe/%5BProject%5D-Data-Pages?node-id=1-19057&t=2xJDhcYuEBp25Rqb-0)
- [OWID iframe sandbox](https://86jv8h.csb.app/#h9)
- [OWID iframe sandbox](https://35h2sn.csb.app/) (datapage only, for performance)

Notes:
- the JSON file is named after the ID of the variable it describes
- ⚠️ the JSON file contains publishing information at its top. This should live in the grapher admin ultimately but is fine for this scale.
- ⚠️ see [caveat about multimetric charts](https://owid.slack.com/archives/C04L4BHSFSB/p1680187372023599)
- validation: fields that are anticipated to be required don't have any safeguards in place. Others do. So only missing required fields might break rendering (and either show an empty space or render a fallback grapher page when an error is thrown). This isn't validation per se, merely an intention that will need more robust handling, e.g. through JSON schema validation.

### CODE REVIEW

**Comment by mlbrgl** [comment]
Swapping what gets rendered based on iframe detection code happens too late, and shows a flash of the full data page in embedded context.

```typescript
export const useIframeDetector = () => {
    const [isEmbedded, setIsEmbedded] = React.useState(false)
    React.useEffect(() => {
        if (window !== window.top) {
            // document.documentElement.classList.add(GRAPHER_IS_IN_IFRAME_CLASS)
            setIsEmbedded(true)
        }
    }, [])
    return isEmbedded
}
``` 
Moving `document.documentElement.classList.add(GRAPHER_IS_IN_IFRAME_CLASS)` back into the `<head>` to apply the `IsInIframe` class doesn't help, as the component swap depends on `isEmbedded`, which happens after render.

Even getting the class from the element itself in `<DataPageContent>` is still happening too late:

```typescript
let isEmbedded = false
    if (
        typeof document !== "undefined" &&
        document.documentElement.classList.contains(GRAPHER_IS_IN_IFRAME_CLASS)
    ) {
        isEmbedded = true
    }
```

What's left is targeting every element on the data page individually, and hiding them through CSS.

File: site/DataPageContent.tsx [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+            <div className="DataPageContent__grapher-for-embed">
+                <GrapherWithFallback
+                    grapher={grapher}
+                    slug={grapherConfig.slug}
+                />
+            </div>
+            <div className="DataPageContent">
+                <div className="header__wrapper wrapper">
+                    <div className="header__left">
+                        <div className="supertitle">DATA</div>
```
Comment (ikesau) [inline site/DataPageContent.tsx:42]: nit: To make things nicer for people who may copy and paste these sections, I prefer writing these with normal title case and using `text-transform: uppercase;`

Comment (ikesau) [reply site/DataPageContent.tsx:42]: Also I recognize this is WIP, but I just want to note for the final version that I'd prefer these classes were written with BEM. `.title` is generic and could end up accidentally getting affected by other components throughout the site.

File: site/GrapherWithFallback.tsx [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+}) => {
+    return slug ? (
+        <div className={cx("GrapherWithFallback", className)}>
+            <>
+                {grapher ? (
+                    <GrapherFigureView grapher={grapher} />
+                ) : (
+                    // Given GrapherFigureView also renders a figure which gets
+                    // the same styling and height, rendering an empty <figure>
+                    // while Grapher is instanciated probably helps with
```
Comment (ikesau) [inline site/GrapherWithFallback.tsx:24]: instantiated*

this is just to have an element to be styled while Grapher loads in, right?

**Review by ikesau — APPROVED** [review]
Nice!

Ignoring the missing mobile styles and UX animations for this WIP stage, the overall architecture of this implementation raises no objection from me 🙂

File: baker/GrapherBaker.tsx [hunk 3]
```diff
@@ -43,18 +43,67 @@ import {
...
     const post = postSlug ? await getPostBySlug(postSlug) : undefined
     const relatedCharts =
         post && isWordpressDBEnabled
             ? await getRelatedCharts(post.id)
             : undefined
-    const relatedArticles =
-        grapher.id && isWordpressAPIEnabled
-            ? await getRelatedArticles(grapher.id)
-            : undefined
+    const relatedArticles = undefined
```
[the remark this answers is not in the document]
Comment (mlbrgl) [reply baker/GrapherBaker.tsx:106]: Good catch, thanks!

File: baker/GrapherBaker.tsx [hunk 4]
```diff
@@ -230,42 +273,12 @@ export const bakeAllPublishedChartsVariableDataAndMetadata = async (
         checksumsDir,
     }))
 
-    if (MAX_NUM_BAKE_PROCESSES == 1) {
```
[the remark this answers is not in the document]
Comment (mlbrgl) [reply baker/GrapherBaker.tsx]: I think I merged from master right in the middle of an experiment @Marigold was running and updated soon after. So my branch still carried those changes and showed them in the diff.

File: site/DataPage.tsx [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+                <meta property="og:image:width" content={imageWidth} />
+                <meta property="og:image:height" content={imageHeight} />
+                <IFrameDetector />
+                <noscript>
+                    <style>{`
+                    figure[data-grapher-src] { display: none !important; }
+                `}</style>
+                </noscript>
+                {flatten(
+                    variableIds.map((variableId) =>
```
[the remark this answers is not in the document]
Comment (mlbrgl) [reply site/DataPage.tsx]: yep, thanks!

File: site/DataPageContent.tsx (hunk 1 above)
Comment (mlbrgl) [reply site/DataPageContent.tsx]: yes, I'm relying on `.title` not being defined at the top level indeed. I was roughly doing BEM down the tree until I felt the extra verbosity wasn't justified. But maybe simpler to just go all the way all the time.

> To make things nicer for people who may copy and paste these sections

@ikesau Not sure about the workflow you're referring to here, and how "DATA" vs "Data" affects it. Turns out, `overline-black-caps` already does the uppercase tranform anyway, but curious still.

**Pull request merged**